### ISSUE TITLE
CSP: form-action and redirects

### ISSUE THREAD

**Issue by mikewest** [w3c/webappsec-csp#8]
_From @ptoomey3 on September 23, 2015 0:12_

I just wanted to open an issue to get your thoughts on `form-action` with respect to redirects. We have been working on deploying `form-action` and have run into a few scenarios that all boil down to:

* Perform some action by doing a `POST` to `self`
* Based on request params/backend state, redirect the user to another site

One big use case of this is OAuth, where we have users submit a form to authorize access to their account. After the `POST` we redirect the user back to the OAuth application to complete the OAuth dance.

However, for these kinds of scenarios to work with `form-action` we would either have to:

* Determine where we plan to redirect the user for each OAuth authorization so that we can add an appropriate value for `form-action`.
* Use something like `meta refresh` to redirect the user after authorizing (since this doesn't count as a form submit)

I can see why one might want to limit method preserving redirection with a 307, but I couldn't think of much risk with allowing 302 style redirects. I guess there is some value in avoiding the redirection, though it feels somewhat orthogonal to the immediate risk associated with accidentally submitting form contents to an untrusted site. Anyway, I was just curious what your thoughts were on the topic. 

/cc @mastahyeti @gregose @oreoshake

_Copied from original issue: w3c/webappsec#482_

**Comment by mikewest** [comment]
Hrm. I think it would be inconsistent with the rest of CSP to differentiate between types of redirection, but you're right to suggest that the risk of a 302 vs 307 is substantially lower for the specific case that `form-action` is meant to deal with. Certainly something worth talking about for CSP3.

Setting label and milestone appropriately.

**Comment by anforowicz** [comment]
FWIW, I think that

A) enforcing form-action on http redirects doesn't offer significant security improvements (protecting against 2.3.3 below is not possible in general;  protecting against 2.3.2 and 2.3.1 doesn't seem important to me).

B) enforcing form-action on http redirects means that the CSP-protected page depends on what should be an implementation details of the server handling the form (i.e. a server receiving the POST data should be free to start responding with an extra, intermediate redirect hop *without* risking breaking some of its clients [some of whom might be using form-action CSP directive]).


I think it is okay for some CSP directives to pay attention to redirects and for some directives (like form-action) to ignore redirects (i.e. only check the origin of the first http request).  I think the comparison between frame-src and form-action below might make it clearer why.

frame-src:

1.1. I assume that frame-src directive mainly protects the parent frame against the threat of embedding a subframe from an unintended origin.

1.2. To protect against the threat above, in theory only the final URI could be verified by CSP .

1.3. This is kind of obvious, but let me also point out that frame-src is enforced for the final URI of both server redirects and client redirects (e.g. via meta refresh and/or window.location).

1.4. frame-src doesn't look at URIs of frames opened in another window (e.g. because the anchor in a subframe had a target attribute OR was shift-clicked). Ignoring other windows seems okay, because the parent-subframe relationship is broken when opening a link in a new window (e.g. the new frame cannot try to trick the user into thinking it is a part of the parent frame's UI). OTOH, in some scenarios window.opener is still populated (e.g. when clicking an anchor with a target attribute, but not when shift-clicking an anchor) so the newly opened frame can still reach into the original parent.

form-action:

2.1. I assume that form-action directive mainly protects form data against the threat of being disclosed to an unintended origin (for example when a hidden form data inside the frame contains an origin-bound secret used for authentication).

2.2. To protect against the threat above, it is important to start CSP enforcement from the very first http request (for example - verifying only the origin of the final redirect destination would be wrong, because the form data could be already disclosed to an unintended origin in the first http request).

2.3. It might be desirable to protect form data against disclosure happening after the first http request. We might want to consider the following threats (personally, I see protection against these threats as much lower priority than protection against the threat outlined in 2.2 above):

2.3.1. Threat of disclosing the form-data to intermediate server redirect destinations (e.g. disclosure to b.com if the redirect chain looks like a.com -> b.com -> c.com). We might want to consider different behavior depending on the type of a server redirect (e.g. http 307 and 308 responses preserve POST data, while other types of redirects don't).

2.3.2. Threat of disclosing the form-data to destinations reached via client redirect (e.g. redirect triggered via meta refresh and/or window.location).

2.3.3. Threat of disclosing the form-data via means other than redirection (e.g. tricking the http server from the original form action into emailing the form data to an unintended origin).

2.4. form-action needs to enforced even if the form action is opened another window (e.g. because the form element had a target attribute OR because the submit button was shift-clicked). This is because the protected resource is the form data (and the form data will be disclosed to the origin at form action URI regardless of whether the request happens in a new window or not).

**Comment by ptoomey3** [comment]
Just ran into another version of this again and thought I'd post here for some context. 

We recently implemented a feature on GitHub.com that allows an organization to require external auth using SAML (in addition to their github auth) to access organization resources. The way this is implemented is that in the event that your external auth session has timed out, we show a modal dialog to let you know you need to do the external auth dance again. This modal is sensibly implemented using a partial that is fetched on demand and shoved into calling page's DOM. This makes sense since this modal might be needed on any page throughout github.com that is associated with a given organization. 

Once the the user clicks "auth with external provider" we do a `POST` to github.com to generate the URL plus params needed to redirect to the external auth provider to do the SAML dance. But, because of CSP, this is kind of quirky to implement. We can't set the additional `form-action` for the auth provider when the modal is fetched via XHR, since the actual `POST` occurs from the page that fetched the modal, and hence that calling page's CSP is the one that rules. So, we will likely be forced to do something such as:

* Fetch modal
* Have form for modal `POST` to `/initiate-external-auth` and NOT attempt a redirect to the external auth provider. Instead, append the auth provider for the organization to `form-action` and render a github.com response interstitial.
* Have that interstitial page do an auto-POST back to github.com that then does the the redirect to the external auth provider. Or,have that interstitial directly do a `GET` to the external auth provider URL. 

It is, on average, pretty tricky for a calling page to know that the endpoint it is `POST`(ing) to might redirect and where it might want to redirect.

**Comment by iquito** [comment]
`form-action` may be the most obscure source for possible problems with a CSP directive - I never would have thought a 302 redirect would still enforce the `form-action` domains, because at that point it is no longer part of the form processing (the user is redirected to a different site with different parameters, nothing of the original form submission is still present in the form, except if specifically specified by the server-side application).

To scrutinize 307 redirects with `form-action` makes sense, but doing it for 302 redirects is confusing for developers and will probably lead to `form-action` being set to very permissive settings in many cases. If the goal is to make it easy to deploy sensible CSP directives it would be good to specifically state that `form-action` should not be applied to 302/303 redirects after form submission.

**Comment by JurRutten** [comment]
Lost hours on this, indeed a confusing and useless scenario.

**Comment by Changaco** [comment]
I've added a warning to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/form-action about the unclear scope and inconsistent browser implementations of this directive.

**Comment by ThrawnCA** [comment]
@anforowicz Re 2.1: Don't forget risks like phishing by injecting markup to assemble a fake login form that will post credentials to the attacker. `form-action` could hinder that by forbidding the fake form from calling home (but that doesn't require any restrictions on redirects).

**Comment by csuhta** [comment]
I agree with others that `form-action` blocking redirects issued from the server is an extremely surprising behavior, and I recently encountered it when implementing an OAuth provider.

Regardless of the final outcome, one of the behaviors should be explicitly defined in CSP3 (is it already somewhere?) 

The current situation where Chrome/Safari block redirects and Firefox does not makes deploying `form-action` very tricky.

**Comment by foolip** [comment]
@mikewest we seem to have run into this problem in the wild with https://github.com/webcompat/webcompat.com/issues/2557, ironically in a project about web compat/interop :)

The error that we saw was: `Refused to send form data to 'http://webcompat.com/' because it violates the following Content Security Policy directive: "form-action 'self'".`

Is this issue at the core of this?

**Comment by andypaicu** [comment]
Reviving this bug. Would everyone be happy if `form-action` applied to to 307/308 redirects but not to other redirects? Does this seem reasonable?

I'm not an expert but I assume if a request goes through a 307 > 302 > 307 redirect chain, it will not send POST the form data to the second 307 redirect (or am I wrong?). But if the redirect chain is 307 > 307 > 307 it will send the POST form data (right?).

Assuming the previous paragraph is correct, `form-action` should apply to all requests that only have 307/308 redirects in their redirect chain.

**Comment by iquito** [comment]
Restricting it to only 307/308 would completely solve it, from my perspective, as 99.9% of current redirects in applications will be 301/302 (with no CSP risk whatsoever), and whenever you do a 307/308 it will most likely be a conscious decision with known security implications - it should not be a surprise there.

**Comment by ptoomey3** [comment]
I think this would resolve the issue for us. We are now up to roughly 4-5 places in our codebase where we do a meta-redirect/JS redirect instead of a 302 just to avoid the current `form-action` behavior. So, a change like this would be much appreciated :smile:

**Comment by Changaco** [comment]
Considering that `form-action` seems to apply regardless of the submission method (`GET` or `POST`), it wouldn't make much sense for it to react differently to different response codes because they change or preserve that HTTP method.

If a server has an open redirect that preserves the query string, then a 302 is as dangerous as a 307, because in that situation they both allow leaking data to third parties. So, if the

**Comment by iquito** [comment]
@Changaco: If you do a GET from a form to a valid `form-action` (like the same domain usually), and then do a 301 or 302 redirect to a different website, that redirect has nothing to do with the form data - no matter if it was GET or POST, because the GET data is not passed along in a redirect, and the POST data isn't either. `form-action` should only apply to where the form is directly going with the form data (so the `action` in the form element, or a 307/308 redirect where POST data is passed along), not redirects after the form submission has been fully processed.

**Comment by Changaco** [comment]
@iquito If a server has an [open redirect vulnerability](https://www.owasp.org/index.php/Unvalidated_Redirects_and_Forwards_Cheat_Sheet) then a 302 response to a form submission can result in form data being transmitted to a third party.

For example if a page in the `example.org` website has a vulnerability that allows an attacker to manipulate form submission, and the website also has an open redirect vulnerability (e.g. `https://example.org/redirect?url=https://malicious.com/$form-data` returns a 302 with `Location: https://malicious.com/$form-data`), then the attacker can get the form data.

If the open redirect vulnerability is really bad and preserves the querystring, then the attacker only needs to be able to manipulate the `method` and `action` attributes of a form to obtain the data, by setting them to `GET` and `https://example.org/redirect?url=https://malicious.com/` respectively, which will result in a `GET` request to `https://malicious.com/?$form-data`.

It seems to me that the "strict" implementation of `form-action` in Chrome is meant to provide protection against such scenarios involving open redirects. Client-side protections against server-side vulnerabilities is a big part of what CSP is, so the "strict" interpretation of `form-action` may be surprising and annoying, but it does make sense.

**Comment by iquito** [comment]
@Changaco If you make a GET request to `https://example.org/redirect?url=https://malicious.com/` the url parameter needs to be urlencoded for it to work, and any form parameters are added with `&` and are not part of the url parameter to which the redirect happens. So:

Form submit to `https://example.org/redirect?url=https%3A%2F%2Fmalicious.com%2F` and add form data `key=value&key2=value` would lead to `https://example.org/redirect?url=https%3A%2F%2Fmalicious.com%2F&key=value&key2=value2` , leading to the following variables being defined in the server-side script:

url = https://malicious.com/
key = value
key2 = value2

Then the script makes a 302 redirect to `https://malicious.com/`. How do you get it to pass the form data?  Is there any documentation of this vulnerability? Because this is not a topic on the open redirect vulnerability page you linked to.

**Comment by iquito** [comment]
@Changaco At that point you have multiple super heavy vulnerabilities to an extent that I don't think you will find a real world example (especially none using a CSP at the same time). Parsing the query string yourself is not a thing in most server side languages - for example in PHP you would never parse your own query string, and this kind of vulnerability could never work. You likely also need access to the HTML and possibly JS code on the page, which needs other vulnerabilities. At that point the `form-action` in CSP is not relevant anymore, because somebody has clearly taken over the website.

On the other hand, the current implementation of `form-action` as implemented in browsers leads to unexpected bugs in most applications, if `form-action` is heavily restricted (as it should be, usually to the current domain). This will most likely lead developers to compensate and therefore to a too permissive `form-action` (including any domains the website could ever possibly redirect to), or to no restriction at all, therefore lessening security for all websites and reducing the impact or importance of `form-action`.

**Comment by ptoomey3** [comment]
> On the other hand, the current implementation of form-action as implemented in browsers leads to unexpected bugs in most applications, if form-action is heavily restricted (as it should be, usually to the current domain). This will most likely lead developers to compensate and therefore to a too permissive form-action (including any domains the website could ever possibly redirect to), or to no restriction at all, therefore lessening security for all websites and reducing the impact or importance of form-action.

I tend to agree. CSP was never meant to be a holistic solution to every single content exfiltration avenue available to an attacker. A tightly scoped `form-action` is handy for multiple reasons that aren't just related to content exfiltration protection. But, as it exists today, many sites are forced through lots of hoops and/or forced to disable the directive.

**Comment by iquito** [comment]
All the more reason to change it soon. If the spec says that form-action does not apply to 301/302 redirects, then browsers will most likely adapt fairly quickly.

It might even be possible to add something like `strict-dynamic` used in `script-src` to `form-action` for progressive enhancement - basically disabling `form-action` for old browsers, like having `form-action * 'redirect-possible' 'self'`, where `redirect-possible` tells new browsers to disregard the catch-all `*` while old browsers have no restrictions. That would fix the problem completely if included in CSP 3.

**Comment by Changaco** [comment]
> having `form-action * 'redirect-possible' 'self'`, where `redirect-possible` tells new browsers to disregard the catch-all `*` while old browsers have no restrictions.

That's an interesting idea, but its drawback is that it sacrifices long-term cleanliness for short-term gain.

> If the spec says that form-action does not apply to 301/302 redirects

I still think that special treatment for 301/302 isn't a good idea. Instead the spec could say that only the URL of the first form submission request should be checked, not the subsequent redirects, either by default or when `form-action` contains `'unsafe-redirects'` or whatever.

**Comment by ThrawnCA** [comment]
> I still think that special treatment for 301/302 isn't a good idea.

IMO the "open redirect that preserves the query string" you've raised is an example of the server shooting itself in the foot, and outside the scope of what CSP can protect it from.

A 302 is not meant to preserve the original request contents. If the server deliberately re-adds those contents, then it has sabotaged itself beyond what a client-side policy should try to fix.

**Comment by iquito** [comment]
> That's an interesting idea, but its drawback is that it sacrifices long-term cleanliness for short-term gain.

"Cleanliness" is not really a big issue (in my opinion), and it is not long-term - as soon as all browsers move on it can be removed, instead of having to wait for browsers to move on until a feature can be used. You are arguing both points now - first that any change will not be usable for years, and then that any workarounds will not be clean. One has to choose one.
 
> I still think that special treatment for 301/302 isn't a good idea. Instead the spec could say that only the URL of the first form submission request should be checked, not the subsequent redirects, either by default or when `form-action` contains `'unsafe-redirects'` or whatever.

I think you are right in that the special treatment should be for the 307/308 redirects: Redirect URLs should specifically be checked against `form-action` if a form is submitted via POST and the response gives a 307/308 redirect, or as long as 307/308 redirects are returned in that scenario. Every other scenario in terms of redirect or response code (or a POST request not coming from a form) should not be hampered by `form-action`.

**Comment by andypaicu** [comment]
> I think the "strict" interpretation of `form-action`, the one currently implemented in Chrome which blocks all redirects to foreign domains regardless of the HTTP method, should be the default, and an `'unsafe-redirect'` keyword should be created to allow reducing the scope of the `form-action` check to the original URL the form is submitted to only. This is simple, flexible, and consistent with the rest of CSP.

I don't think this is enough in terms of backwards compatibility. With only this you can not write a policy with `'unsafe-redirect'` that also does not break in older versions of CSP which means that you can not really use it and you have to resort to client-side redirects. That's why a separate `'disregards-wildcards'` keyword does not make much sense to me since you always have to use it when you `'unsafe-redirects'` otherwise your policy simply breaks form submissions of agents implements the strict version of `form-action`.

Also I suspect there is no need to prefix it with `unsafe`(if we decide for a keyword). I think this makes the case that form submissions are pretty safe to not need to check redirects.

**Comment by Changaco** [comment]
The `'disregard-wildcard'` keyword is meant to be temporary, it would become unnecessary when the browsers that don't support `'unsafe-redirect'` become negligible. Until then the `'disregard-wildcard'` and `'unsafe-redirect'` keywords would indeed always be used together by websites that don't implement the client-side redirect approach. I see nothing wrong with that.

To summarize, with my proposal a website developer would have 3 choices:

- implement client-side redirects and use a strict `form-action 'self'`
- implement client-side redirects and use a less strict `form-action 'self' 'unsafe-redirect'`, with the aim of dropping the client-side redirects once they become unnecessary
- use `form-action 'self' * 'disregard-wildcard' 'unsafe-redirect'` to avoid the need for client-side redirects, at the cost of lessened security in current browsers

After several years the third choice would become safer and would be simplified to the same directive as the second choice (`form-action 'self' 'unsafe-redirect'`). The `'disregard-wildcard'` keyword would then be marked as obsolete in the CSP spec.

**Comment by iquito** [comment]
`'unsafe-redirect'` does not make sense - this is the point of this whole issue: redirects are not unsafe at all, except for 307/308 after a form POST. In every other scenario it is always safe to redirect, as the redirect has nothing to do with the form submission.

At the same time, doing a 307/308 redirect after a form POST should never avoid scrunity - you are re-POST-ing a form to another URL, which should conform to `form-action` as the form is redirected as-is. I do not see any reason to specifically allow this, because at that point you can just put the new URL in `form-action` instead of allowing a 307/308 to ANY existing URL.

Having something like `form-action * 'disregard-wildcard' 'self'` and still enforcing `form-action` when a form POST then encounters a 307/308 would solve all current problems without incurring new ones.

**Comment by Changaco** [comment]
(I've just realized that the [current draft of CSP3](https://w3c.github.io/webappsec-csp/) already has a provisional `'unsafe-allow-redirects'` keyword for the new `navigation-to` directive.)

**Comment by ThrawnCA** [comment]
> I think it would be incoherent for `form-action` to allow 302s by default if `navigation-to` blocks them.

Not really? I mean, 302s drop the contents of a form, so it makes sense not to restrict them with `form-action`.

**Comment by annevk** [comment]
The problem isn't just with `POST`, right? There's also exfiltration. And if the problem was just with `POST` it shouldn't block a `GET` form at all, which seems like a bad idea?

**Comment by Changaco** [comment]
If `form-action` doesn't block 302s, then the usefulness of the `navigation-to` directive is severely reduced, because it would always be possible to bypass it with a `GET` form, since `form-action` supersedes `navigation-to`. For this reason, `form-action` needs to be as secure as `navigation-to` by default, i.e. it needs to block all non-matching redirects, no matter what their HTTP status is or how deep they are in the redirect chain, like Chrome currently does.

Yes, this strict interpretation of `form-action` makes it more difficult to use dynamic redirects in response to form submissions, but is that really so bad? There is a relatively simple solution to this problem that multiple websites already use (client-side redirects), and the `'unsafe-allow-redirects'` keyword could provide another solution that is even easier to implement.

**Comment by Changaco** [comment]
> For this reason, `form-action` needs to be as secure as `navigation-to` by default

Actually that's not true, it could lax by default and have a keyword to make it strict, but that would be inconsistent with how the other CSP directives work (safest by default, with `unsafe` keywords to make them more lenient).

**Comment by andypaicu** [comment]
`'unsafe-allow-redirects'` actually does something different. It will check the last response in the redirect chain (and we would like to check the first one).

Overall, especially looking at how `form-actions` is related to `navigate-to`, I believe them aligning with each other in terms of behavior is reasonable for CSP3. Keep in mind that `navigate-to` will defer to `form-action` for form initiated navigations so keeping consistent behavior between the two seems very important.

I think the current strict behavior is the way to go for the near future and a detailed note or even a section should be added to make it clear to developers that this is the behavior. I'm hoping that the biggest issue here is that the behavior can be unexpected at first glance and perhaps a simple explanation is sufficient. (That and the fact that browsers are not interoperable here).

**Comment by Changaco** [comment]
> `'unsafe-allow-redirects'` actually does something different. It will check the last response in the redirect chain (and we would like to check the first one).

Indeed, I hadn't realized that. Why is it designed that way? Shouldn't it check the first URL instead of the last one, in order to protect against intermediate redirects to malicious sites?

**Comment by andypaicu** [comment]
For the use cases that people seemed to care around `navigate-to` it made more sense like this. The main one is around ads and ensuring that you land on the proper page (even if you don't know through what sort of click-counting hops you go through) eventually.

**Comment by Changaco** [comment]
Okay, then we need a separate keyword that means "only check the first request URL, follow all redirects after that".

**Comment by Changaco** [comment]
Maybe the current `'unsafe-allow-redirects'` should be renamed `'unsafe-detours'`, and a new `'unsafe-redirects'` should be created.

**Comment by leeoniya** [comment]
just ran into this trying to implement an Affirm checkout which doesn't rely on their bloated client-side script. tried doing a POST to route on own server that returns a 302 to an Affirm url with a checkout token. i suspect i'm about to hit the same problem with Paypal's similar flow.

+1 for allowing 301s & 302s returned by self as a result of form submits.

**Comment by thecopy** [comment]
This results in _less_ security in our OAuth2.0 solution. Where we have to remove `form-action` completely to get the redirect to work in Chrome.

I recommend this to allow redirects here. At least for `HTTP 303`.

**Comment by ptoomey3** [comment]
Given there is a fairly broadly deployed implementation in the wild for a number of years that implements the exact behavior a lot of folks want/expect (Firefox), might it be worth looking at changing this? It probably crops up about once a quarter internally and forces folks to jump through awkward hoops (not a great sign when you get to the point of having developed an internal meta redirect library/view for folks to use for this 😂). If the default behavior changed, it would seemingly be "backward compatible". I guess the one edge case is folks who explicitly didn't want the behavior...but those folks are already in a "bad spot" since there is nothing they can do today to "protect" Firefox users.

**Comment by annevk** [comment]
I consider this a bug in Firefox, to be clear. I'd support adding `unsafe-allow-redirects` to `form-action`, but I don't think `form-action` should have different semantics from `navigate-to` and ignoring redirects without opt-in doesn't seem sound.

**Comment by ptoomey3** [comment]
> I'd support adding unsafe-allow-redirects

That would also work

**Comment by mikewest** [comment]
Sorry this fell off everyone's radar. @andypaicu isn't working on CSP these days, @antosart is a better contact person for Chrome.

I share @annevk's feelings about the default behavior, but adding an opt-in `'unsafe-whatever'` flag that would check the initial hop only isn't unreasonable. It is unfortunate that this is the opposite of the flag we originally designed for `navigate-to`, which had a very different threat in mind (malvertising, as opposed to exfiltration). But, happily(?), no one shipped `navigate-to`, so renaming its flag is likely a reasonable thing to do.

I'm hard-pressed to say that this is a priority for Chrome, but perhaps @antosart can find some time to sketch something out?

**Comment by thecopy** [comment]
`unsafe-allow-redirects` sounds good to me as well.

**Comment by antosart** [comment]
To bring in yet another point of view, I think it makes sense to have a conceptual distinction between
  - 'form-action', which restricts pages where form data can be submitted
  - 'navigate-to', which restricts pages which the current document is allowed to navigate to

According to this logic, for `<form>` I would just apply both until a 302/303 redirect, after which I would only apply 'navigate-to'. I personally find the current logic "'navigate-to' applies to forms only if there is no 'form-action'" confusing and not particularly helpful.

That said, adding an opt-in flag also seems reasonable. I would question whether it needs the `unsafe-` prefix, since it does not seem unsafe at all and I believe it would just scare developers.

**Comment by iquito** [comment]
`unsafe-allow-redirects` seems reasonable for 307/308 redirects (although there it would be *really* unsafe), not for 301/302 redirects, where I still do not see any reason to declare it unsafe.

I have been using very strict CSP settings for all my applications in the last 4 years, except for `form-action`, which is always set to `*` (with a comment linking to this thread), because often I will have a form submission and then redirect to a payment provider, or a form submission to then initiate an OAuth authorization flow. I cannot risk even adding additional domains to `form-action`, because they might change (or not be completely knowable in advance) and break an application unexpectedly.

In addition to that, adding more domains to `form-action` would make my application *less* secure compared to just `'self'`, as my application should never POST directly to these domains, it only redirects to them in very specific circumstances without any of the form data. I definitely would like to prevent my website to POST the login form to any domain except `'self'`.

I would very much welcome a better implementation of `form-action` - it is the only part of CSP that seems completely broken to me.

**Comment by Changaco** [comment]
I don't know why this discussion is suddenly so active again, but I want to remind everyone that there is a workaround.

When redirecting to another domain, you can modify your server code to:

- return a `200` HTTP response code instead of a `302`
- return a `Refresh` HTTP header instead of a `Location` header, for example: 
  `Location: https://example.com/foo` → `Refresh: 0;url=https://example.com/foo`
- fill the HTTP response body with a simple HTML page informing the user that they're being redirected

**Comment by iquito** [comment]
@Changaco: The Refresh HTTP header is not defined in any standard and is only used as a convention by browsers. Switching away from an official HTTP standard to something only browsers support for legacy reasons seems like a terrible idea. Many HTTP libraries do not support the Refresh header at all - why would they, it is not part of the HTTP spec.

**Comment by jub0bs** [comment]
Just my two cents: I recently realised that an XSLeak attack can leverage `form-action`'s enforcement on HTTP redirects to detect a cross-site redirect that requires a cookie marked `SameSite=Lax`. More details at https://xsleaks.dev/docs/attacks/navigations/#csp-violations

**Comment by Sora2455** [comment]
> Note that this approach is unviable in Firefox (contrary to Chromium-based browsers) because form-action doesn’t block redirects after a form submission in that browser.

Does this mean that resolving this ambiguity is now a security and/or privacy issue?

**Comment by jub0bs** [comment]
@Sora2455 Yes, at least insofar as I was able to exploit this technique on a prominent messaging app for de-anonymisation purposes. Blog post incoming.

**Comment by mig5** [comment]
> In addition to that, adding more domains to form-action would make my application less secure compared to just 'self', as my application should never POST directly to these domains, it only redirects to them in very specific circumstances without any of the form data. I definitely would like to prevent my website to POST the login form to any domain except 'self'.

I can add even one more addition, in the case of OAuth/OIDC: having to list all the domains that one might've issued an OAuth2.0 client for, in the form-action, actually also introduces an information disclosure vulnerability, in that examining the CSP at the OP can leak implicit information about what domains may be RPs. Obviously any individual RP gives this information away about itself by invoking the OAuth flow, but maybe we don't want to expose that information about *all* RPs at the OP side.

Finally, this is a semantic point, but I don't like the idea of having to use any attribute referred to as 'unsafe' given that the OAuth2/OIDC specs say that the redirect URL [MUST](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) be validated as being that which is associated with the OAuth2.0 client itself, before redirecting the browser to it. An implementation of the OP that does not do this, has a security vulnerability. Therefore, at least in the case of OAuth2/OIDC, correctly implementing the spec is not at all something that a browser should consider 'unsafe' - it lacks context as to whether other safety measures are in place. I think 'ignore-redirect' is a nicer term.

**Comment by roberto-cisternino** [comment]
Is the SSL-offloading interpreted as redirect due to the protocol change ?
My CSP form-action fails for this reason I think...

**Comment by Sora2455** [comment]
@marek22k You can use a client-side redirect instead of server-side, but that involves an extra delay in redirecting and more complexity in your application.

**Comment by marek22k** [comment]
Mhh, that's dumb. I thought I could maybe use an additional CSP header or something like `allow-redirect 'self';`?!

**Comment by Sora2455** [comment]
@marek22k Essentially the problem (as I understand it, I do not work for a browser vendor) is that the form-action directive is designed to prevent form submissions from being sent to random websites (e.g. because someone injected a form into your page).

"But hang on, redirects don't cause the form data to be sent to the page we're redirecting to? So a redirect after a form submission can't cause form data to be sent to random websites?"

They can if it's a [307 Temporary Redirect](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/307)/[308 Permanent Redirect](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/308) status code redirect.

"Well, why don't we just block form redirects to unlisted destinations on a 307/308, but allow them on 301/303 redirects?"

I believe the objection here is that no other CSP directive depends on the exact response code, so this could be a great big "Gotcha!" for developers. (Much like the current behaviour).

"Well... alright then, why not add a 'allow-redirects' keyword to the directive or something?"

This thread has been mostly arguing about the exact syntax of such a keyword and what form it would take. e.g. Should it have the 'unsafe' keyword? If the directive isn't present, should redirects be allowed or denied?

"But this issue was lodged over 6 years ago! Why haven't they agreed on a solution since then?!"

Again, don't work for a browser vendor... but I get the impression that this isn't really a priority for them to solve.

**Comment by roberto-cisternino** [comment]
I believe form-action is not usable successfully cross-browser... report-to can be used to obtain some more info from a browser.
Btw, I noted that once csp is enabled the browser behaviour is strict more then ever... things like Google tag manager are not working anymore without specific settings... and CSP fails to provide more info if ther is an error... CSP handbook seems to be not mature enouth for production... we are playing the testing game !

**Comment by Sora2455** [comment]
@marek22k That sounds like a browser error - if you can make a minimal test case, and it only happens in some browsers, then you might want to lodge a bug against that browser or browsers.

**Comment by marek22k** [comment]
My webapp could be used as an example:
Goto https://dn42-bgplookup.mk16.de/ and enter `12345678` at `ASN goto`.
CSP: `default-src 'none'; base-uri 'none'; form-action 'self'; frame-ancestors 'none'; style-src 'self'; style-src-elem 'self'; img-src 'self';`
Chromium shows `Refused to send form data to 'https://dn42-bgplookup.mk16.de/asn?asn=12345678' because it violates the following Content Security Policy directive: "form-action 'self'".` in the console, Firefox redirects the request as requested.

Where would I have to report this for Chromium?

**Comment by fredericDelaporte** [comment]
marek22k case has some other cause on its server side (protocol downgrade, so if redirects to other than self are refused, it should be blocked), as this can now be seen on its chromium bug report. So, these comments are actually irrelevant and should be hidden as "Outdated" or "Off-topic" in my opinion, included my own comment.
